packages/ckeditor5-ui/src/textarea/textareaview.ts

```diff
@@ -0,0 +1,211 @@
+
+		this.extendTemplate( {
+			attributes: {
+				class: [ 'ck-textarea' ],
+				style: {
+					height: bind.to( '_height', height => height ? toPx( height ) : null ),
+					resize: bind.to( 'resize' )
+				},
+				rows: bind.to( 'minRows' )
+			}
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public override render(): void {
+		super.render();
+
+		this.on( 'input', () => {
+			this._updateAutoGrowHeight( true );
+		} );
+
+		this.on( 'change:value', () => {
+			// The content needs to be updated by the browser after the value is changed. It takes a few ms.
+			global.window.requestAnimationFrame( () => this._updateAutoGrowHeight() );
+		} );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public override reset(): void {
+		super.reset();
+
+		this._updateAutoGrowHeight();
+	}
+
+	/**
+	 * Updates the {@link #_height} of the view depending on {@link #minRows}, {@link #maxRows}, and the current content size.
+	 *
+	 * **Note**: This method overrides manual resize done by the user using a handle. It's a known bug.
+	 */
+	private _updateAutoGrowHeight( shouldScroll?: boolean ): void {
+		const viewElement = this.element!;
+		const singleLineContentClone = getTextareaElementClone( viewElement, '1' );
+		const fullTextValueClone = getTextareaElementClone( viewElement, viewElement.value );
+		const singleLineContentStyles = singleLineContentClone.ownerDocument.defaultView!.getComputedStyle( singleLineContentClone );
+
+		const verticalPaddings = parseFloat( singleLineContentStyles.paddingTop ) + parseFloat( singleLineContentStyles.paddingBottom );
+		const borders = getBorderWidths( singleLineContentClone );
+		const lineHeight = parseFloat( singleLineContentStyles.lineHeight );
+		const verticalBorder = borders.top + borders.bottom;
+
+		const singleLineAreaDefaultHeight = new Rect( singleLineContentClone ).height;
+		const numberOfLines = Math.round( ( fullTextValueClone.scrollHeight - verticalPaddings ) / lineHeight );
+
+		const maxHeight = this.maxRows * lineHeight + verticalPaddings + verticalBorder;
+
+		// There's a --ck-ui-component-min-height CSS custom property that enforces min height of the component.
+		// This min-height is relevant only when there's one line of text. Other than that, we can rely on line-height.
+		const minHeight = numberOfLines === 1 ? singleLineAreaDefaultHeight : this.minRows * lineHeight + verticalPaddings + verticalBorder;
+
+		// The size of textarea is controlled by height style instead of rows attribute because event though it is
+		// a more complex solution, it is immune to the layout textarea has been rendered in (gird, flex).
+		this._height = Math.min(
+			Math.max(
+				Math.max( numberOfLines, this.minRows ) * lineHeight + verticalPaddings + verticalBorder,
+				minHeight
+			),
+			maxHeight
+		);
+
+		if ( shouldScroll ) {
+			viewElement.scrollTop = viewElement.scrollHeight;
+		}
+
+		singleLineContentClone.remove();
+		fullTextValueClone.remove();
+
+		this.fire<TextareaViewAutoGrowEvent>( 'autoGrow' );
```

Comment:
Yes, it's consumed in another project. This is a bit sketchy, though because the aim of this event was to know when anything happens to the content of textarea, either by user input or by declarative input change. The auto-grow logic acts on both events (even if the height didn't change) so it was convenient to fire an event here.

I think we could move this `fire()` to the right place (somewhere in the constructor) but I'm at a loss when it comes to naming this kind of event. 

Context: we use this event to determine whether the scrollbar showed up or disappeared. This is not necessarily associated with the change of height (e.g. pressing enter when a number of lines === maxRows does not change height but introduces a scrollbar).

Reply:
Decision: Let's name the even update and fire it in related listeners that may change the geometry of the component (`input`, `change:value` + `reset()`)